File: learnLeetCode/Day8/Part2/three_sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Day 8 — Part 2: 3Sum
// Approach: Sort + two-pointers, skip duplicates.
// Time: O(n^2), Space: O(1) extra (excluding output)
vector<vector<int>> threeSum(vector<int> nums) {
    sort(nums.begin(), nums.end());
    int n = nums.size();
    vector<vector<int>> res;
    for (int i = 0; i < n; ++i) {
        if (i > 0 && nums[i] == nums[i-1]) continue; // skip duplicate anchor
        if (nums[i] > 0) break; // remaining will be positive
        int l = i + 1, r = n - 1;
        while (l < r) {
            long long s = 1LL*nums[i] + nums[l] + nums[r];
            if (s == 0) {
                res.push_back({nums[i], nums[l], nums[r]});
                ++l; --r;
                while (l < r && nums[l] == nums[l-1]) ++l;
                while (l < r && nums[r] == nums[r+1]) --r;
            } else if (s < 0) {
                ++l;
            } else {
                --r;
            }
        }
    }
    return res;
}

int main() {
    vector<int> nums = {-1,0,1,2,-1,-4};
    auto ans = threeSum(nums);
    // print result in any order
    cout << "Triplets:" << "\n";
    for (const auto& t : ans) {
        cout << "[" << t[0] << "," << t[1] << "," << t[2] << "]\n";
    }

    // basic checks
    // We can't assert order; just check size is 2 for given example
    if (threeSum({-1,0,1,2,-1,-4}).size() != 2) return 1;
    if (threeSum({0,0,0}) != vector<vector<int>>{{0,0,0}}) return 1;
    if (threeSum({}).size() != 0) return 1; // empty ok
    return 0;
}

File: Day8/Part2/three_sum.py
"""
Day 8 — Part 2: 3Sum

Find all unique triplets that sum to zero.
Approach: Sort + O(n^2) two-pointers, skip duplicates for i, left, and right.
Time: O(n^2), Space: O(1) extra (excluding output).
"""
from typing import List


def three_sum(nums: List[int]) -> List[List[int]]:
    nums.sort()
    n = len(nums)
    res: List[List[int]] = []

    for i in range(n):
        if i > 0 and nums[i] == nums[i - 1]:
            continue  # skip duplicate anchor
        if nums[i] > 0:
            break     # no further triplets possible

        l, r = i + 1, n - 1
        while l < r:
            s = nums[i] + nums[l] + nums[r]
            if s == 0:
                res.append([nums[i], nums[l], nums[r]])
                # move left and right to next different values
                l += 1
                while l < r and nums[l] == nums[l - 1]:
                    l += 1
                r -= 1
                while l < r and nums[r] == nums[r + 1]:
                    r -= 1
            elif s < 0:
                l += 1
            else:
                r -= 1
    return res


if __name__ == "__main__":
    assert three_sum([-1,0,1,2,-1,-4]) in (
        [[-1,-1,2],[-1,0,1]],
        [[-1,0,1],[-1,-1,2]],
    )
    assert three_sum([]) == []
    assert three_sum([0]) == []
    assert three_sum([0,0,0]) == [[0,0,0]]
    print("Day8 Part2 three_sum.py: all tests passed")